protocol/statetracker/updaters/provider_availability_updater.go

```diff
@@ -0,0 +1,87 @@
+package updaters
+
+import (
+	"context"
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/lavanet/lava/v3/protocol/lavasession"
+	"github.com/lavanet/lava/v3/protocol/metrics"
+	"github.com/lavanet/lava/v3/x/pairing/types"
+	"google.golang.org/grpc"
+)
+
+type ProviderPairingStatusStateQueryInf interface {
+	Providers(ctx context.Context, in *types.QueryProvidersRequest, opts ...grpc.CallOption) (*types.QueryProvidersResponse, error)
+}
+
+type AvailabilityStatus uint64
+
+const (
+	AVAILABLE AvailabilityStatus = iota
+	FROZEN
+)
+
+type ProviderAvailabilityUpdater struct {
+	lock                sync.RWMutex
+	latestEpoch         uint64
+	pairingQueryClient  ProviderPairingStatusStateQueryInf
+	metricsManager      *metrics.ProviderMetricsManager
+	clientCtx           client.Context
+	rpcProviderEndpoint *lavasession.RPCProviderEndpoint
+}
+
+func NewProviderAvailabilityUpdater(
+	stateQuery ProviderPairingStatusStateQueryInf,
+	rpcProviderEndpoint *lavasession.RPCProviderEndpoint,
+	clientCtx client.Context,
+	metricsManager *metrics.ProviderMetricsManager,
+) *ProviderAvailabilityUpdater {
+	return &ProviderAvailabilityUpdater{
+		pairingQueryClient:  stateQuery,
+		clientCtx:           clientCtx,
+		rpcProviderEndpoint: rpcProviderEndpoint,
+		metricsManager:      metricsManager,
+	}
+}
+
+func (pau *ProviderAvailabilityUpdater) UpdateEpoch(epoch uint64) {
+	go pau.runProviderAvailabilityUpdate(epoch)
+}
+
+func (pau *ProviderAvailabilityUpdater) runProviderAvailabilityUpdate(epoch uint64) {
+	pau.lock.Lock()
+	defer pau.lock.Unlock()
+
+	if epoch <= pau.latestEpoch {
+		return
+	}
+	ctx := context.Background()
```

Comment:
Never use context background when calling stuff you can hang forever. 
use a context with timeout of 10 seconds (define in a const)